session/service.go

```diff
@@ -0,0 +1,121 @@
+package session
+
+import (
+	"context"
+	"time"
+
+	"github.com/influxdata/influxdb/v2"
+	"github.com/influxdata/influxdb/v2/rand"
+	"github.com/influxdata/influxdb/v2/snowflake"
+)
+
+type Service struct {
+	store         *Storage
+	userService   influxdb.UserService
+	urmService    influxdb.UserResourceMappingService
+	sessionLength time.Duration
+
+	idGen    influxdb.IDGenerator
+	tokenGen influxdb.TokenGenerator
+}
+
+func NewService(store *Storage, userService influxdb.UserService, urmService influxdb.UserResourceMappingService, sessionLength time.Duration) *Service {
+	return &Service{
+		store:         store,
+		userService:   userService,
+		urmService:    urmService,
+		sessionLength: sessionLength,
+		idGen:         snowflake.NewIDGenerator(),
+		tokenGen:      rand.NewTokenGenerator(64),
+	}
+}
+
+func (s *Service) FindSession(ctx context.Context, key string) (*influxdb.Session, error) {
+	return s.store.FindSessionByKey(ctx, key)
+}
+
+func (s *Service) ExpireSession(ctx context.Context, key string) error {
+	session, err := s.store.FindSessionByKey(ctx, key)
+	if err != nil {
+		return err
+	}
+	return s.store.DeleteSession(ctx, session.ID)
+}
+
+func (s *Service) CreateSession(ctx context.Context, user string) (*influxdb.Session, error) {
+	// lookup user
+	u, err := s.userService.FindUser(ctx, influxdb.UserFilter{
+		Name: &user,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	mappings, _, err := s.urmService.FindUserResourceMappings(ctx, influxdb.UserResourceMappingFilter{UserID: u.ID}, influxdb.FindOptions{Limit: 100})
+	if err != nil {
+		return nil, err
+	}
+
+	permissions, err := permissionSet(mappings, u.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	// account for paginated results
+	if len(mappings) == 100 {
+		// if we got 100 mappings we probably need to pull more pages
+		for i := len(mappings); len(mappings) > 0; {
+			mappings, _, err = s.urmService.FindUserResourceMappings(ctx, influxdb.UserResourceMappingFilter{UserID: u.ID}, influxdb.FindOptions{Offset: i, Limit: 100})
+			if err != nil {
+				return nil, err
+			}
+			pms, err := permissionSet(mappings, u.ID)
+			if err != nil {
+				return nil, err
+			}
+			permissions = append(permissions, pms...)
+			i += len(mappings)
+		}
+	}
+
+	token, err := s.tokenGen.Token()
+	if err != nil {
+		return nil, err
+	}
+
+	session := &influxdb.Session{
+		ID:          s.idGen.ID(),
+		Key:         token,
+		CreatedAt:   time.Now(),
+		ExpiresAt:   time.Now().Add(s.sessionLength),
```

Comment:
This is a take it or leave it nit, but if we use the same value for now here then it will always be precisely sessionLength between created and expires.